src/Hocon/HoconObject.cs

```diff
@@ -0,0 +1,363 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="HoconImmutableObject.cs" company="Akka.NET Project">
+//      Copyright (C) 2013 - 2020 .NET Foundation <https://github.com/akkadotnet/hocon>
+// </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+
+namespace Hocon
+{
+    public class HoconObject :
+        HoconElement,
+        IReadOnlyDictionary<string, HoconElement>,
+        IDictionary<string, HoconElement>
+    {
+        public static readonly HoconObject Empty = new HoconObject();
+
+        protected ImmutableSortedDictionary<string, HoconElement> Fields;
+
+        protected HoconObject()
+        {
+            Fields = ImmutableSortedDictionary<string, HoconElement>.Empty;
+        }
+
+        protected HoconObject(HoconElement source)
+        {
+            if (!(source is HoconObject other))
+                throw new HoconException($"Could not create HoconObject from {source.GetType()}");
+
+            Fields = other.ToImmutableSortedDictionary();
+        }
+
+        protected HoconObject(IDictionary<string, HoconElement> fields)
+        {
+            Fields = fields.ToImmutableSortedDictionary();
+        }
+
+        protected HoconObject(IReadOnlyDictionary<string, HoconElement> fields)
+        {
+            Fields = fields.ToImmutableSortedDictionary();
+        }
+
+        public override HoconType Type => HoconType.Object;
+
+        public override HoconElement this[HoconPath path] => GetValue(path);
+
+        public new HoconElement this[string path] {
+            get => GetValue(path);
+            set => throw new InvalidOperationException("HoconObject is a read only Dictionary.");
```

Comment:
Can we avoid this by just not inheriting from `IDictionary` - leave it as `IReadOnlyDictionary` only? Or did you need this for testing?

Reply:
`IDictionary` interface is needed for backward compatibility, I've blocked all of the write behaviour